In [1]:
!pip install xlrd #read xls
!pip install openpyxl #read xlsx
import pandas as pd
import requests
import time

Defaulting to user installation because normal site-packages is not writeable
--- Logging error ---
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pip/_internal/utils/logging.py", line 177, in emit
    self.console.print(renderable, overflow="ignore", crop=False, style=style)
  File "/usr/local/lib/python3.8/dist-packages/pip/_vendor/rich/console.py", line 1752, in print
    extend(render(renderable, render_options))
  File "/usr/local/lib/python3.8/dist-packages/pip/_vendor/rich/console.py", line 1390, in render
    for render_output in iter_render:
  File "/usr/local/lib/python3.8/dist-packages/pip/_internal/utils/logging.py", line 134, in __rich_console__
    for line in lines:
  File "/usr/local/lib/python3.8/dist-packages/pip/_vendor/rich/segment.py", line 245, in split_lines
    for segment in segments:
  File "/usr/local/lib/python3.8/dist-packages/pip/_vendor/rich/console.py", line 1368, in render
    renderable = rich_cast(renderable)
  File 

In [2]:
request_path = "http://127.0.0.1:8000/api/pesquisa/iniciacao_cientifica"

In [3]:
path = '/home/ricardo/Downloads/CPq_Dados_Atena_Mai_2022 - divulgação.xlsx'

In [4]:
listadpt = ['FLA', 'FLP', 'FLF', 'FLH', 'FLC', 'FLM', 'FLO', 'FLL', 'FSL', 'FLT', 'FLG']

listadptext = ['Antropologia', 'Ciência Política', 'Filosofia', 'História', 'Letras Clássicas e Vernáculas', 
                'Letras Modernas', 'Letras Orientais', 'Lingüística', 'Sociologia', 
                'Teoria Literária e Literatura Comparada', 'Geografia']





In [37]:
dfDiff = pd.DataFrame(columns={'codproj','COD PROJETO', 'ano_proj', 'ANO PROJETO', 'nome_discente', 
                'ALUNO', 'status_projeto', 'SITUAÇÃO PROJETO', 'titulo_pesquisa', 
                'TITULO', 'nome_departamento', 'DEPARTAMENTO'}) 

for dpt, dptext in zip(listadpt, listadptext):
    for ano in range(2016, 2021):
        parametros = {
        'ano': ano,
        'departamento': dpt,
        'departamento_ext': dptext,
        'export': 'false',
        'tipo': 'anual'
        }

        
        time.sleep(1)
        com = requests.get(request_path+"?bolsa=true", params= parametros)
        time.sleep(0.5)
        sem = requests.get(request_path+"?bolsa=false", params= parametros)
        
        try:    
            bolsa = pd.DataFrame(com.json())
            sem_bolsa = pd.DataFrame(sem.json())
        except:
            print(parametros)
            print(com.status_code) 
            print('\n')
            continue
            
        portal = pd.concat([bolsa, sem_bolsa], ignore_index=True)

        cpq = pd.read_excel(path)

        cpq = cpq.loc[(cpq['DEPARTAMENTO'] == parametros['departamento_ext']) & (cpq['ANO PROJETO'] == int(parametros['ano'])) ]
        cpq = cpq.reset_index(drop=True)

        portal = portal[['codproj', 'ano_proj', 'nome_discente', 'titulo_pesquisa', 'status_projeto', 'nome_departamento']]
        cpq =  cpq[['COD PROJETO', 'ANO PROJETO', 'ALUNO', 'TITULO', 'SITUAÇÃO PROJETO', 'DEPARTAMENTO']]

        cpq = cpq.astype({'COD PROJETO' : str, 'ANO PROJETO' : str})
        portal = portal.astype({'codproj' : str, 'ano_proj' : str})

        df = pd.merge(
        portal, 
        cpq, 
        how='outer', 
        left_on=['codproj', 'ano_proj'], 
        right_on=['COD PROJETO', 'ANO PROJETO']
                        )

        column_names = ['codproj','COD PROJETO', 'ano_proj', 'ANO PROJETO', 'nome_discente', 
            'ALUNO', 'status_projeto', 'SITUAÇÃO PROJETO', 'titulo_pesquisa', 
            'TITULO', 'nome_departamento', 'DEPARTAMENTO']

        df = df.reindex(columns=column_names)
        df = df.sort_values(by=['ALUNO'], ascending=True)
        df = df.reset_index(drop=True)
       
        #df = df.loc[df['nome_discente'].isnull() | df['ALUNO'].isnull()]
        
        dfDiff=dfDiff.append( [df] ) # deprecated
        dfDiff = pd.concat([dfDiff, df])

          

In [32]:
column_names = ['codproj','COD PROJETO', 'ano_proj', 'ANO PROJETO', 'nome_discente', 
            'ALUNO', 'status_projeto', 'SITUAÇÃO PROJETO', 'titulo_pesquisa', 
            'TITULO', 'nome_departamento', 'DEPARTAMENTO']

dfDiff = dfDiff.reindex(columns=column_names)
dfDiff = dfDiff.sort_values(by=['ALUNO'], ascending=True)
dfDiff = dfDiff.reset_index(drop=True)
dfDiff

codproj  COD PROJETO ano_proj ANO PROJETO  \
0      NaN         2784      NaN        2020   
1      NaN          514      NaN        2020   
2      NaN         3064      NaN        2020   
3      NaN  1276 e 3910      NaN        2018   
4      NaN         2313      NaN        2020   
5      NaN    306 e 998      NaN        2020   
6      NaN         2693      NaN        2020   
7      NaN         2671      NaN        2020   
8       46          NaN     2016         NaN   
9     3910          NaN     2018         NaN   
10    2576          NaN     2018         NaN   
11     306          NaN     2020         NaN   
12    1348          NaN     2016         NaN   

                         nome_discente  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   
5                                  NaN   
6                                  NaN   
7                                  NaN   
8                Daniel Tamayo de Lima   
9   Gabriel Rodrigues Sanches Cordeiro   
10              Guaraí Pereira Machado   
11          João Cândido Cartocci Maia   
12   Carina Ferreira de Holanda Mendes   

                                         ALUNO status_projeto  \
0                       Bruno Batista de Sousa            NaN   
1   Bruno Fontana Nishiyama Bernardes Ferreira            NaN   
2                       Bruno Mariano Horemans            NaN   
3           Gabriel Rodrigues Sanches Cordeiro            NaN   
4                             Gabrielly Borges            NaN   
5                   João Cândido Cartocci Maia            NaN   
6             Lucas Augusto dos Santos Camargo            NaN   
7                       Vitor Sposito de Faria            NaN   
8                                          NaN    Transferido   
9                                          NaN       Aprovado   
10                                         NaN       Aprovado   
11                                         NaN       Aprovado   
12                                         NaN    Transferido   

           SITUAÇÃO PROJETO  \
0                     Ativo   
1                     Ativo   
2                     Ativo   
3   Transferido (2018-3909)   
4                     Ativo   
5                  Aprovado   
6         Ativo (2020-1022)   
7                 Concluido   
8                       NaN   
9                       NaN   
10                      NaN   
11                      NaN   
12                      NaN   

                                      titulo_pesquisa  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                     Projeto de Iniciação Científica   
9   Fome e Abastecimento na Região Parisiense Entr...   
10  UMA CULTURA TEATRAL DE RESISTÊNCIA:\nTHE FREED...   
11  Amar-amaro: desejo, política e poesia no prime...   
12  Análise das transformações dos setores censitá...   

                                               TITULO  \
0   Compilação, tradução e análise de lemas latino...   
1                                     Platão e o Mito   
2     História e Memória na ficção de Valêncio Xavier   
3   Fome e abastecimento na região parisiense entr...   
4   Ou isto ou aquilo: discussão do estatuto de li...   
5   Amar-amaro: desejo, política e poesia no prime...   
6   Método automatizado para mapeamento de habitat...   
7   Cultive apenas borboletas no seu estômago. Die...   
8                                                 NaN   
9                         

In [6]:
dfDiff.to_csv('ics diferenças dados - cpq (2).csv', encoding='utf-8', index=False)

PASSO A PASSO ====>>

In [13]:
pd.reset_option("display.max_rows", "display.max_columns")

In [8]:
parametros = {
    'ano': "2019",
    'departamento': 'FLM',
    'departamento_ext': 'Letras Modernas',
    'export': 'false',
    'tipo': 'anual'
}

In [9]:
request_path = "http://127.0.0.1:8000/api/pesquisa/iniciacao_cientifica"

In [11]:
com = requests.get(request_path+"?bolsa=true", params= parametros)
sem = requests.get(request_path+"?bolsa=false", params= parametros)

In [12]:
bolsa = pd.DataFrame(com.json())
sem_bolsa = pd.DataFrame(sem.json())

In [13]:
portal = pd.concat([bolsa, sem_bolsa], ignore_index=True)

In [14]:
print("n bolsa: ",len(bolsa), "n sem bolsa: ", len(sem_bolsa), ' | Total: ', len(bolsa) + len(sem_bolsa))

n bolsa:  1 n sem bolsa:  24  | Total:  25


In [15]:
path = '/home/ricardo/Downloads/CPq_Dados_Atena_Mai_2022 - divulgação.xlsx'

In [16]:
cpq = pd.read_excel(path)

In [17]:
cpq.DEPARTAMENTO.unique()

array(['Antropologia', 'Ciência Política', 'Filosofia', 'Geografia',
       'História', 'Letras Clássicas e Vernáculas', 'Letras Modernas',
       'Letras Orientais', 'Lingüística', 'Sociologia',
       'Teoria Literária e Literatura Comparada'], dtype=object)

In [18]:
cpq.loc[(cpq['DEPARTAMENTO'] == parametros['departamento_ext'])]

ANO PROJETO COD PROJETO UNIDADE     DEPARTAMENTO DATA INICIO PROJETO  \
109          2016        2404   FFLCH  Letras Modernas          2016-09-01   
110          2016         278   FFLCH  Letras Modernas          2016-11-01   
111          2016        3592   FFLCH  Letras Modernas          2017-02-01   
112          2016        2002   FFLCH  Letras Modernas          2016-11-01   
113          2016        3572   FFLCH  Letras Modernas          2016-12-01   
...           ...         ...     ...              ...                 ...   
1400         2021         335   FFLCH  Letras Modernas          2021-04-01   
1493         2022         314   FFLCH  Letras Modernas          2022-03-14   
1494         2022         233   FFLCH  Letras Modernas          2021-11-27   
1495         2022         165   FFLCH  Letras Modernas          2021-12-01   
1496         2022         218   FFLCH  Letras Modernas          2021-11-25   

      MESES DURAÇÃO DATA FIM PROJETO  CODIGO AREA CONHECIMENTO  \
109       12.133333       31/08/2017                  80106005   
110       12.133333       31/10/2017                  80208002   
111       12.133333       31/01/2018                  80202004   
112       12.133333       31/10/2017                  80202004   
113       12.133333       30/11/2017                  80202004   
...             ...              ...                       ...   
1400      12.133333       31/03/2022                  80208002   
1493      12.133333       13/03/2023                  80205003   
1494      12.133333       26/11/2022                  80202004   
1495      12.133333       30/11/2022                  80208002   
1496      12.133333       24/11/2022                  80202004   

                                      AREA CONHECIMENTO  \
109    Lingüística, Letras e Artes-Lingüística Aplicada   
110   Lingüística, Letras e Artes-Literaturas Estran...   
111   Lingüística, Letras e Artes-Línguas Estrangeir...   
112   Lingüística, Letras e Artes-Línguas Estrangeir...   
113   Lingüística, Letras e Artes-Línguas Estrangeir...   
...                                                 ...   
1400  Lingüística, Letras e Artes-Literaturas Estran...   
1493       Lingüística, Letras e Artes-Teoria Literária   
1494  Lingüística, Letras e Artes-Línguas Estrangeir...   
1495  Lingüística, Letras e Artes-Literaturas Estran...   
1496  Lingüística, Letras e Artes-Línguas Estrangeir...   

                                          PALAVRA CHAVE  ...  \
109        língua alemã; linguística; sintaxe do alemão  ...   
110                         conto; literatura; história  ...   
111   língua italiana;material didático;cultura ital...  ...   
112         ensino de espanhol; currículo; Ensino Médio  ...   
113   Breakfast at Tifany's;Audrey Hepburn;Truman Ca...  ...   
...                                                 ...  ...   
1400  adaptação; obra literária occitana; album musical  ...   
1493   Georg Lukács; Mikhail Bakhtin; teoria do romance  ...   
1494          língua alemã; livro didático; vocabulário  ...   
1495  poesia chilena; Soledad Fariña; poesia contemp...  ...   
1496        Germanística; revista acadêmica; editoração  ...   

     AREA PRIORITARIA MCTIC  Unnamed: 32 Unnamed: 33 Unnamed: 34  Unnamed: 35  \
109                     NaN          NaN         NaN         NaN          NaN   
110                     NaN          NaN         NaN         NaN          NaN   
111                     NaN          NaN         NaN         NaN          NaN   
112                     NaN          NaN         NaN         NaN          NaN   
113                     NaN          NaN         NaN         NaN          NaN   
...                     ...          ...         ...         ...          ...   
1400                    NaN          NaN         NaN         NaN          NaN   
1493                    NaN          NaN         NaN         NaN          NaN   
1494                    NaN          NaN         NaN         NaN     

In [19]:
cpq = cpq.loc[(cpq['DEPARTAMENTO'] == parametros['departamento_ext']) & (cpq['ANO PROJETO'] == int(parametros['ano'])) ]
cpq = cpq.reset_index(drop=True)

In [20]:
print("Total Portal Dados: ", len(portal), '. Total CPQ: ', len(cpq))

Total Portal Dados:  25 . Total CPQ:  25


In [21]:
portal = portal[['codproj', 'ano_proj', 'nome_discente', 'titulo_pesquisa', 'status_projeto', 'nome_departamento']]
cpq =  cpq[['COD PROJETO', 'ANO PROJETO', 'ALUNO', 'TITULO', 'SITUAÇÃO PROJETO', 'DEPARTAMENTO']]

In [22]:
cpq.dtypes

COD PROJETO         object
ANO PROJETO          int64
ALUNO               object
TITULO              object
SITUAÇÃO PROJETO    object
DEPARTAMENTO        object
dtype: object

In [23]:
cpq = cpq.astype({'COD PROJETO' : str, 'ANO PROJETO' : str})


In [24]:
portal.dtypes

codproj               int64
ano_proj              int64
nome_discente        object
titulo_pesquisa      object
status_projeto       object
nome_departamento    object
dtype: object

In [25]:
portal = portal.astype({'codproj' : str, 'ano_proj' : str})


In [26]:
portal.dtypes

codproj              object
ano_proj             object
nome_discente        object
titulo_pesquisa      object
status_projeto       object
nome_departamento    object
dtype: object

In [27]:
df = pd.merge(
    portal, 
    cpq, 
    how='outer', 
    left_on=['codproj', 'ano_proj'], 
    right_on=['COD PROJETO', 'ANO PROJETO']
)

In [28]:
column_names = ['codproj','COD PROJETO', 'ano_proj', 'ANO PROJETO', 'nome_discente', 
                'ALUNO', 'status_projeto', 'SITUAÇÃO PROJETO', 'titulo_pesquisa', 
                'TITULO', 'nome_departamento', 'DEPARTAMENTO']

df = df.reindex(columns=column_names)
df = df.sort_values(by=['ALUNO'], ascending=True)
df = df.reset_index(drop=True)

In [29]:
print("Total Portal Dados: ", len(portal), '. Total CPQ: ', len(cpq))

Total Portal Dados:  25 . Total CPQ:  25


In [30]:
df

codproj COD PROJETO ano_proj ANO PROJETO  \
0      143         143     2019        2019   
1     3056        3056     2019        2019   
2     2122        2122     2019        2019   
3      468         468     2019        2019   
4     3730        3730     2019        2019   
5      318         318     2019        2019   
6     2234        2234     2019        2019   
7     3416        3416     2019        2019   
8      411         411     2019        2019   
9     2960        2960     2019        2019   
10    2980        2980     2019        2019   
11    2684        2684     2019        2019   
12     331         331     2019        2019   
13    1821        1821     2019        2019   
14    3230        3230     2019        2019   
15    3231        3231     2019        2019   
16     342         342     2019        2019   
17    3516        3516     2019        2019   
18    3631        3631     2019        2019   
19    2811        2811     2019        2019   
20    3586        3586     2019        2019   
21      24          24     2019        2019   
22    2952        2952     2019        2019   
23     374         374     2019        2019   
24    3843        3843     2019        2019   

                            nome_discente  \
0          Ana Carolina de Assis Perfeito   
1                Ana Clara Neves Silveira   
2                       Ayne Silva Soares   
3                 Beatriz de Souza Santos   
4                 Camila Marcucci Schmidt   
5                Carolina Martins Pedroso   
6                  Cesar Henrique Fachina   
7                   Dayane Manfrere Alves   
8               Douglas Patriota Ferreira   
9                Gabriela Hikari Tomizuka   
10         Gabriella Fernandes Rampinelli   
11          Giuliana Cerchiari de Andrade   
12                  Helena de Luna Mendes   
13                 Henrique Elias Roveran   
14                 Joao Pedro Costa Jorge   
15                   Julio Inhasz de Melo   
16                  Larissa da Silva Rosa   
17       Leonardo Baroni Alves dos Santos   
18           Mariana de Carvalho Oliveira   
19                     Marina Evangelista   
20       Pedro Oswaldo Horta Martins Pena   
21                    Rafaela Xavier Rëgo   
22  Sofia Helena Arneiro Lourenco Barbosa   
23            Tainá Alves Domingues Brito   
24               Tamara de Farias Cervera   

                                    ALUNO status_projeto SITUAÇÃO PROJETO  \
0          Ana Carolina de Assis Perfeito       Aprovado         Aprovado   
1                Ana Clara Neves Silveira       Aprovado         Aprovado   
2                       Ayne Silva Soares       Aprovado         Aprovado   
3                 Beatriz de Souza Santos       Aprovado         Aprovado   
4                 Camila Marcucci Schmidt       Aprovado         Aprovado   
5                Carolina Martins Pedroso       Aprovado         Aprovado   
6                  Cesar Henrique Fachina       Aprovado         Aprovado   
7                   Dayane Manfrere Alves       Aprovado         Aprovado   
8               Douglas Patriota Ferreira       Aprovado         Aprovado   
9                Gabriela Hikari Tomizuka       Aprovado         Aprovado   
10         Gabriella Fernandes Rampinelli       Aprovado         Aprovado   
11          Giuliana Cerchiari de Andrade       Aprovado         Aprovado   
12                  Helena de Luna Mendes       Aprovado         Aprovado   
13                 Henrique Elias Roveran       Aprovado         Aprovado   
14                 Joao Pedro Costa Jorge       Aprovado         Aprovado   
15                   Julio Inhasz de Melo       Aprovado         Aprovado   
16                  Larissa da Silva Rosa       Aprovado         Aprovado   
17       Leonardo Baroni Alves dos Santos       Aprovado         Aprovado   
18           Mariana de Carvalho Oliveira       Aprovado         Aprovado   
19                     Marina Evangelista       Aprovado         Apro

In [57]:
df.loc[df['nome_discente'].isnull() | df['ALUNO'].isnull()]

Empty DataFrame
Columns: [codproj, COD PROJETO, ano_proj, ANO PROJETO, nome_discente, ALUNO, status_projeto, SITUAÇÃO PROJETO, titulo_pesquisa, TITULO, nome_departamento, DEPARTAMENTO]
Index: []

In [58]:
df['nome_discente'].duplicated().any() # EXISTE ALUNO REPETIDO?

False

In [115]:
df.loc[df['nome_discente'].duplicated()] # & df['nome_discente'] is not None

Empty DataFrame
Columns: [codproj, COD PROJETO, ano_proj, ANO PROJETO, nome_discente, ALUNO, status_projeto, SITUAÇÃO PROJETO, titulo_pesquisa, TITULO, DEPARTAMENTO]
Index: []

In [116]:
obs = '''\n\nHistória 2021: 
            3336, 2821,3620 -> Projeto na planilha cpq ativo, mas agr no replicadoe stá cancelado (Projeto cancelado segundo solicitação do orientador.)
            962 -> Status transferido, substituição bolsista, vigencia > 30 dias
           Em geral está tudo certo: poral 32, cpq 33'''

In [117]:
nome_arquivo = "observações sobre ic (equalizando dados com o cpq)"
try:
    arquivo = open(nome_arquivo, 'r+')#edita arquivo
except FileNotFoundError:
    arquivo = open(nome_arquivo, 'w+')#cria o arquivo
    
arquivo.readlines()
arquivo.writelines(obs)
arquivo.close()





História 2018: Há dois projetos com status transferido e que consta na planilha da cpq 
        (1276 - Gabriel Sanches, 1165 - Rafael Gonçalves)
         O 1276 ele é transferido pois a bolsa mudou, de PIBIC foi para FAPESP.
         O 1165 foi transferido pq mudou o bolsista, mas deve contar pq a vigencia é de 2 meses.
         Mas no geral Portal: 40 CPQ: 41 (O CPQ junta os projetos do gabeial 1276 e 3910), a diferença é do Raphael que não consta no portal, pois o status é transferido
         PROJ COM VIGENCIA MENOR QUE 30 DIAS DEVE SER REMOVIDO
História 2020: Tem um projeto cancelado recentemente no replicado, mas que na planilha da cris aparecia como concluído. Como agora está cancelado então vamos desconsiderar.
           Em geral está tudo certo: poral 32, cpq 33
História 2021: 
            3336, 2821, -> Projeto na planilha cpq ativo, mas agr no replicadoe stá cancelado (Projeto cancelado segundo solicitação do orientador.)
            962 -> Status transferido, substituição bolsista, vigencia > 30 dias
           Em geral está tudo certo: poral 32, cpq 33


In [27]:
%pwd


'/home/ricardo'

In [332]:
#https://towardsdatascience.com/calculating-string-similarity-in-python-276e18a7d33a
!pip install python-Levenshtein

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.5/50.5 kB 8.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for python-Levenshtein: filename=python_Levenshtein-0.12.2-cp38-cp38-linux_x86_64.whl size=166088 sha256=60b5f47f73b306d6dce254dfd605ef9b1f8cecda84698042e154974a38f12ed7
  Stored in directory: /home/ricardo/.cache/pip/wheels/d7/0c/76/042b46eb0df65c3ccd0338f791210c55ab79d209bcc269e2c7
Successfully built python-Levenshtein


In [333]:
import Levenshtein

In [343]:
Levenshtein.distance("Fome e abastecimento na região parisiense entre os séculos VII e XI", 'Fome e Abastecimento na Região Parisiense Entre os Séculos VII e XI')

5

In [344]:
Levenshtein.distance("Amar-amaro: desejo, política e poesia no primeiro Drummond.", "Amar-amaro: desejo, política e poesia no primeiro Drummond")

1